### PR TITLE
[logos] Support multi-line block arguments in `%orig` calls

### LINKED ISSUE #6: Logos errors out on calling %orig with an inline block with multiple lines.
When hooking a method which accepts a block as a paramter, Logos dislikes this form of calling the original implementation:

```
%hook AwesomeClass

- (void)AwesomeMethod:(void (^)(void))awesomeBlock
{
    %orig(^{
        NSLog(@"Hello");
    }); 
}

%end
```

However, this works:

```
%hook AwesomeClass

- (void)AwesomeMethod:(void (^)(void))awesomeBlock
{
    %orig(^{ NSLog(@"Hello"); }); 
}

%end
```

and putting the block in a variable, works too, obviously.

Logos errors out with this error: "error: missing closing parenthesis"

### PR DESCRIPTION
**Pull request by EthanArbuckle**
What does this implement/fix? Explain your changes.
---------------------------------------------------
This PR adds the ability to specify inline blocks spanning multiple lines as arguments when calling `%orig(...)`, which was previously not supported.

Example `%orig(...)` usages that are now supported:

```
%orig(^{
    NSLog(@"foo");
});
```

```
%orig(arg1, arg2, ^(NSString *blockArg1) {
    NSLog(@"block arg: %@", blockArg1);
});
```

```
- (void)methodWithComplexBlock:(void (^)(NSString *, NSError *))completion {
    %orig(^(NSString *result, NSError *error) {
        if (error) {
            NSLog(@"Error: %@", error);
        }
        completion(result, error);
    });
}
```

Does this close any currently open issues?
------------------------------------------
This PR closes an issue that's been open for 10y: [\[#6\] "Logos errors out on calling %orig with an inline block with multiple lines."](https://github.com/theos/logos/issues/6)

Any relevant logs, error output, etc?
-------------------------------------
N/A

Any other comments?
-------------------
* This maintains backwards compatibility with existing Logos syntax. Developers do not need to modify their projects/code to accommodate this change.
* Using logos directives (`%orig`, `%log`, etc.) inside of a block passed to `%orig(...)` is currently **not supported**.
    * Attempting to do so will result in the following error: **Tweak.x:4: error: Logos cannot be used within arguments to %orig**.

Example of unsupported usage:

```
%orig(arg1, arg2, ^(NSString *blockArg1) {
    %orig(arg1, arg2, nil);
});

Tweak.x:4: error: Logos cannot be used within arguments to %orig
```

    
Where has this been tested?
---------------------------
**Operating System:** macOS Sonoma, iOS 14, iOS 18, Ubuntu 22.0.4

**Platform:** arm64, x86_64

**Target Platform:** iOS, iOS Simulator

**Toolchain Version:** Xcode 15.4

**SDK Version:** iOS 14, iOS 15, iOS 16.5

Unit Testing
------------

AFAIK, Logos (and Theos) lack official unit testing. This PR changes core behavior of the parser and has the potential to introduce significant breakage; this requires thorough testing to ensure reliability and backwards compatibility. To address this concern, I created a separate branch containing unit tests for my changes as well as some unrelated core behavior, as the project lacks these otherwise.

Coverage of these tests include, but is not limited to:

1. Basic `%orig` functionality to ensure backwards compatibility.
2. New multi-line block parsing capabilities.
3. Error handling for invalid syntax.
4. Edge cases such as complex argument structures nested `%orig` usage.

These tests validate the correctness of the current implementation and provide a safety net for future changes (helping prevent regressions). All of the tests I created can be found [here](https://github.com/EthanArbuckle/logos/tree/unit-tests/tests). 

[Direct link to the tests for `%orig` behavior](https://github.com/EthanArbuckle/logos/blob/unit-tests/tests/test_orig_directive.py).

I have Github Actions setup to run these tests (in my fork) -- they're currently passing: [Latest Github Action run](https://github.com/EthanArbuckle/logos/actions/runs/10354353990/job/28659432852).

### CODE REVIEW

**Review by uroboro — APPROVED** [review]
I finally had time to go over this and understand it.

File: bin/lib/Logos/Generator/Base/Function.pm [hunk 1]
```diff
@@ -194,4 +196,5 @@ sub notSoSmartSplit {
 	return $args;
 }
 
+
```
Comment (L1ghtmann) [inline bin/lib/Logos/Generator/Base/Function.pm:199]: nitpick: extra line

**Review by L1ghtmann — CHANGES_REQUESTED** [review]
Thanks a ton for putting this together! I reviewed this a little while back and all looked good. Can't say anything stands out upon second glance. Thanks for adding the check for Logos within `%orig`. Totally agree that a test suite is in order. It's on the list!

Edit: we'll also probably want to update https://theos.dev/docs/logos-syntax to note that block syntax will be valid

File: bin/lib/Logos/Generator/Base/Function.pm (hunk 1 above)
Comment (EthanArbuckle) [reply bin/lib/Logos/Generator/Base/Function.pm:199]: bruh

File: bin/lib/Logos/Generator/Base/Function.pm [hunk 2]
```diff
@@ -2,6 +2,8 @@ package Logos::Generator::Base::Function;
 use strict;
 use Logos::Generator;
 use Logos::Util;
+use parent qw(Logos::Generator::Base::Common);
+
```
Comment (L1ghtmann) [inline bin/lib/Logos/Generator/Base/Function.pm:6]: image

Comment (EthanArbuckle) [reply bin/lib/Logos/Generator/Base/Function.pm:6]: I would have been thrilled to receive and address any substantive feedback from all reviewers

I’m not coming back to this after months for a whitespace (respectfully)

**Review by L1ghtmann — APPROVED** [review]

**Comment by EthanArbuckle** [comment]
@L1ghtmann Thank you! Feel free to merge (I lack perms)

**Comment by EthanArbuckle** [comment]
This issue (12 years old!) can be closed: https://github.com/theos/logos/issues/6
